accepted/2020/platform-exclusion/platform-exclusion.md

```diff
@@ -0,0 +1,217 @@
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=true, Inherited=false)]
+    public sealed class UnsupportedOSPlatformAttribute : OSPlatformAttribute
+    {
+        public UnsupportedOSPlatformAttribute(string platformName);
+    }
+```
+
+The semantics of this attribute are as follows:
+
+* Marks an assembly, module, or API as unsupported on a given OS platform.
+* The `platformName` can contain no version number, a single version number, or
+  a range:
+    - `browser`. Marks the API as unsupported on all version of `browser`. This
+      will be used for APIs that can't be supported on a particular OS platform.
+    - `ios14.0`. Marks the API as unsupported starting with iOS 14.0. This will
+      be used for OS platform APIs that were removed/disabled.
+    - `windows0.0-10.0.19041`. Marks the API as unsupported for 10.0.19041 and
+      earlier. This will be used for APIs that were unsupported before but can
+      be supported starting with a later version of the OS. This is different
+      from applying `[MinimumOSPlatform("windows10.0.19041")]` in that the
+      presence of this attribute doesn't cause the API to be considered
+      unsupported on other platforms.
+
+### Analyzer behavior
+
+The existing analyzer that handles `[MinimumOSPlatform]` will be modified to
+handle `[UnsupportedOSPlatform]` (previously named `[RemovedInOSPlatform]`)
+with the semantics listed above.
+
+Let's look at a few examples of annotations what the corresponding guard clauses
+are that would prevent the analyzer from flagging usage of the API.
+
+```C#
+// The API is supported everywhere except when running in a browser
+[UnsupportedOSPlatform("browser")]
+public extern void Api();
+
+public void Api_Usage()
+{
+    if (!RuntimeInformation.IsOSPlatform("browser"))
+    {
+        Api();
+    }
+}
+```
+
+```C#
+// On Windows, the API was unsupported up to and including version 10.0.19041.
+// The API is considered supported everywhere else without constraints.
+[UnsupportedOSPlatform("windows0.0-10.0.19041")]
+public extern void Api();
+
+public void Api_Usage()
+{
+    if (!RuntimeInformation.IsOSPlatformOrEarlier("windows10.0.19041"))
+    {
+        Api();
+    }
+}
+```
+
+```C#
+// The API is only supported on iOS. There, it started to be supported in
+// version 12.0 and stopped being supported in version 14.
+[MinimumOSPlatform("iOS12.0")]
+[UnsupportedOSPlatform("ios14.0")]
+public extern void Api();
+
+public void Api_Usage()
+{
+    if (RuntimeInformation.IsOSPlatformOrLater("iOS12.0") &&
+        RuntimeInformation.IsOSPlatformOrEarlier("ios14.0"))
```

Comment:
```suggestion
        RuntimeInformation.IsOSPlatformEarlierThan("ios14.0"))
```

Fixes the logic bug, and uses a method that actually exists.  Win/win :smile:.

Reply:
I've changed this to this:

```C#
    if (RuntimeInformation.IsOSPlatformOrLater("ios12.0") &&
        !RuntimeInformation.IsOSPlatformOrLater("ios14.0"))
    {
        Api();
    }
```

@buyaa-n can we track state like this in the analyzer? You need to compute a lower bound and an upper bound.

* `RuntimeInformation.IsOSPlatformOrLater("iOS12.0")`. You know platform is `isOS` and the minimum is `12.0`
* `!RuntimeInformation.IsOSPlatformOrLater("os14.0")`. You know the platform is either not `iOS` or the version is higher than 14.
* `RuntimeInformation.IsOSPlatformOrLater("iOS12.0") && !RuntimeInformation.IsOSPlatformOrLater("os14.0")`. You know the platform is `iOS` and the version range is `[12.0,14.0)`.

Reply:
It's not supported with the current implementation, but i think we can make that evaluation based on flow analysis result looks `&&` logic is more doable than `||`